pmd-core/pom.xml

```diff
@@ -186,6 +186,12 @@
             <artifactId>system-rules</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```

Comment:
Do we really need this dependency? I think, you added it to use AbstractIterator. Would it be possible to write the RemainingTokensIterator without this?

Reply:
The reason was indeed to be able to use the AbstractIterator. In my opinion, this makes the code much more readable. However, if the Guava dependency is undesirable, it should be possible to rewrite the iterator without needing Guava.

Reply:
Note that the PMD 7 branch has an AbstractIterator class already. I based it off the kotlin stdlib, but it looks like sensibly the same. We could just backport it, I don't think it needs Java 8

https://github.com/pmd/pmd/blob/01c7b12df4916ed56dcedc64f8c4f33793abac1f/pmd-core/src/main/java/net/sourceforge/pmd/internal/util/IteratorUtil.java#L361

Reply:
@oowekyala Good suggestion! I've backported this class and adjusted my implementation to use it.